### PR TITLE
Scaleup AE blocker w/o multiprocessing

### PR DESCRIPTION
**Pull request by sanjibkd**
This pull request has:
+ AE blocker with code optimizations (e.g., pushing down projections, itertuples, caching blocking attribute values in a dictionary)
+ 16 unit test cases for block_tables and block_candset with 100% code coverage
+ 9 benchmark test cases -- 6 for block_tables and 3 for block_candset
+ benchmark datasets
+ no documentation yet (the function signatures will change on adding multiprocessing)

### CODE REVIEW

File: magellan/tests/test_attr_equiv_blocker.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+r_block_attr_3 = 'name'
+l_output_attrs = ['zipcode', 'birth_year']
+r_output_attrs = ['zipcode', 'birth_year']
+l_output_prefix = 'l_'
+r_output_prefix = 'r_'
+
+bogus_attr = 'bogus'
+block_attr_multi = ['zipcode', 'birth_year']
+
+
```
Comment (kvpradap) [inline magellan/tests/test_attr_equiv_blocker.py]: @sanjibkd 

Can you please add testcases checking for invalid input type. For example the input A table is not a dataframe, the block attributes are not of type string, the prefixes and output attrs are not of type string.

You can do sthg like this for checking

    if not isinstance(A, pd.DataFrame):
        logger.error('Input left table is not of type pandas data frame')
        raise AssertionError('Input left table is not of type pandas data frame')

    if not isinstance(l_prefix, six.string_types):
        logger.error('l_prefix is not of type string')
        raise AssertionError('l_prefix is not of type string')

Comment (kvpradap) [inline magellan/tests/test_attr_equiv_blocker.py]: @sanjibkd, the code looks fine to me. Can you pl. add the unit test cases mentioned, commit and push to the branch. 

You dont have to create a pull request. The current pull request will get automatically updated if you push to scaleup_AE_blocker branch.

File: magellan/tests/test_attr_equiv_blocker.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        mg.set_key(B, r_key)
+        ab = mg.AttrEquivalenceBlocker()
+        C = ab.block_tables(A, B, l_block_attr_1, r_block_attr_1, None, [bogus_attr])
+
+    def test_ab_block_candset(self):
+        A = mg.read_csv_metadata(path_for_A)
+        mg.set_key(A, l_key)
+        B = mg.read_csv_metadata(path_for_B)
+        mg.set_key(B, r_key)
+        ab = mg.AttrEquivalenceBlocker()
```
Comment (paulgc) [inline magellan/tests/test_attr_equiv_blocker.py]: All the tests seem to have these following lines,

A = mg.read_csv_metadata(path_for_A)
mg.set_key(A, l_key)
B = mg.read_csv_metadata(path_for_B)
mg.set_key(B, r_key)
ab = mg.AttrEquivalenceBlocker()

Can we put it in a "setup" function ?

File: asv.conf.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    // The project's homepage
+    "project_url": "https://github.com/anhaidgroup/magellan/",
+
+    // The URL or local path of the source code repository for the
+    // project being benchmarked
+    "repo": "https://github.com/anhaidgroup/magellan.git",
+
+    // List of branches to benchmark. If not provided, defaults to "master"
+    // (for git) or "tip" (for mercurial).
+    "branches": ["scaleup_AE_blocker"], // for git
```
Comment (paulgc) [inline asv.conf.json]: @sanjibkd  should this option be changed to default, when pushing code ?

File: magellan/blocker/attr_equiv_blocker.py [hunk 4]
```diff
@@ -150,48 +162,37 @@ def validate_block_attrs(self, ltable, rtable, l_block_attr, r_block_attr):
             r_block_attr = [r_block_attr]
         assert set(r_block_attr).issubset(rtable.columns) is True, 'Right block attribute is not in the right table'
 
-    def output_columns(self, l_key, r_key, col_names, l_output_attrs, r_output_attrs, l_output_prefix, r_output_prefix):
+    def output_columns(self, l_key, r_key, col_names, l_output_attrs,
```
Comment (paulgc) [inline magellan/blocker/attr_equiv_blocker.py]: @sanjibkd  the utility functions should be made private (i.e., function name prefixed by _ )?

File: magellan/tests/test_attr_equiv_blocker.py (hunk 1 above)
Comment (sanjibkd) [inline magellan/tests/test_attr_equiv_blocker.py]: @kvpradap done

File: magellan/tests/test_attr_equiv_blocker.py (hunk 2 above)
Comment (sanjibkd) [inline magellan/tests/test_attr_equiv_blocker.py]: @paulgc done

File: asv.conf.json (hunk 3 above)
Comment (sanjibkd) [inline asv.conf.json]: @paulgc done

File: magellan/blocker/attr_equiv_blocker.py (hunk 4 above)
Comment (sanjibkd) [inline magellan/blocker/attr_equiv_blocker.py]: @paulgc i will rename functions and variables when i do pylinting

**Comment by sanjibkd** [comment]
@kvpradap should i go ahead and merge this pull request (i have made the changes you and @paulgc requested)

**Comment by kvpradap** [comment]
@sanjibkd. I will merge it now.

**Pull request merged**